python/lib/metrics/dmod/metrics/communication.py

```diff
@@ -147,20 +294,131 @@ def verbosity(self) -> Verbosity:
+    def error(self, message: str, exception: Exception = None, verbosity: Verbosity = None, publish: bool = None):
+        if verbosity and self._verbosity < verbosity:
+            return
+
+        if exception and exception.__traceback__:
+            formatted_exception = os.linesep.join(
+                traceback.format_exception(
+                    type(exception),
+                    exception,
+                    exception.__traceback__
+                )
+            )
+            print(formatted_exception, file=sys.stderr)
+        elif exception:
+            message += f" ERROR: {exception}"
+
+        if self.__include_timestamp:
+            timestamp = datetime.now().astimezone().strftime("%Y%m%d %H:%M%z")
+            message = f"[{timestamp}] {message}"
+
+        print(message, file=sys.stderr)
+
+        if publish:
+            self.write(reason="error", data={"error": message})
+
+        # Call every event handler for the 'error' event
+        for handler in self._handlers.get("error", []):
+            handler(message)
+
+    def info(self, message: str, verbosity: Verbosity = None, publish: bool = None):
+        if self.__include_timestamp:
+            timestamp = datetime.now().astimezone().strftime("%Y%m%d %H:%M%z")
+            message = f"[{timestamp}] {message}"
+
+        print(message)
+
+        if publish:
+            self.write(reason="info", data={"info": message})
+
+        # Call every event handler for the 'info' event
+        for handler in self._handlers.get("info", []):
+            handler(message)
+
+    def read_errors(self) -> typing.Iterable[str]:
+        return (message for message in self.__errors)
+
+    def read_info(self) -> typing.Iterable[str]:
+        return (message for message in self.__info)
+
+    def _validate(self) -> typing.Sequence[str]:
+        return []
+
+    def write(self, reason: ReasonToWrite, data: dict):
+        """
+        Writes data to the communicator's channel
+
+        Takes the form of:
+
+        {
+            "event": reason,
+            "time": YYYY-mm-dd HH:MMz,
+            "data": json string
+        }
+
+        Args:
+            reason: The reason for data being written to the channel
+            data: The data to write to the channel; will be converted to a string
+        """
+        message = {
+            "event": reason,
+            "time": datetime.now().astimezone().strftime("%Y%m%d %H:%M%z"),
+            "data": data
+        }
+
+        pprint(message, indent=4)
+
+        try:
+            for handler in self._handlers.get('write', []):
+                handler(message)
+        except:
+            # Leave room for a breakpoint
+            raise
+
+    def read(self) -> typing.Any:
+        return input(self.__read_message)
+
+    def update(self, **kwargs):
+        self.__properties.update(kwargs)
+
+    def __getitem__(self, item):
+        return self.__properties[item]
+
+    def sunset(self, seconds: float = None):
```

Comment:
I'm going to go ahead and remove the sunset function - needed for the redis communicator, but that can be handled over there instead of across the board.